experimental/unicodeset_parser/src/parse.rs

```diff
@@ -0,0 +1,1192 @@
+
+    let mut strings = multi.into_iter().collect::<Vec<_>>();
+    strings.sort();
+    let zerovec = (&strings).into();
+
+    let cpinvlistandstrlist = CodePointInversionListAndStringList::try_from(built_single, zerovec)
+        .map_err(|_| PEK::Internal)?;
+
+    Ok(cpinvlistandstrlist)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    macro_rules! td {
+        () => {
+            icu_testdata::unstable()
+        };
+    }
+
+    const OPTIONS_ANCHOR: UnicodeSetBuilderOptions = UnicodeSetBuilderOptions {
+        dollar_is_anchor: true,
+    };
+    const OPTIONS_NO_ANCHOR: UnicodeSetBuilderOptions = UnicodeSetBuilderOptions {
+        dollar_is_anchor: false,
+    };
+
+    fn assert_set_equality<'a, 'b>(
+        cpinvlistandstrlist: &CodePointInversionListAndStringList<'a>,
+        single: impl Iterator<Item = char>,
+        multi: impl Iterator<Item = &'b str>,
+    ) {
+        let mut it_size = 0;
+        for c in single {
+            it_size += 1;
+            assert!(
+                cpinvlistandstrlist.contains_char(c),
+                "missing char from parsed set: '{}'",
+                c.escape_debug()
+            );
+        }
+        for s in multi {
+            it_size += 1;
+            assert!(
+                cpinvlistandstrlist.contains(s),
+                "missing string from parsed set: \"{}\"",
+                s.escape_debug()
+            );
+        }
+        assert_eq!(cpinvlistandstrlist.size(), it_size);
+    }
+
+    fn assert_is_error_and_message_eq(
+        options: UnicodeSetBuilderOptions,
+        source: &str,
+        expected_err: &str,
+    ) {
+        let result = parse_unstable(source, options, &td!());
+        assert!(result.is_err());
+        let err = result.unwrap_err();
+        assert_eq!(err.fmt_with_source(source).to_string(), expected_err);
+    }
+
+    #[test]
+    fn test_semantics() {
+        let cases: [(_, _, _, Vec<String>); 10] = [
+            (OPTIONS_NO_ANCHOR, "[a]", 'a'..='a', vec![]),
+            (OPTIONS_NO_ANCHOR, r"[\n]", '\n'..='\n', vec![]),
+            (OPTIONS_NO_ANCHOR, "[\\\n]", '\n'..='\n', vec![]),
+            // empty - whitespace is skipped
+            (OPTIONS_NO_ANCHOR, "[\n]", 'b'..='a', vec![]),
+            (OPTIONS_NO_ANCHOR, "[a-z]", 'a'..='z', vec![]),
+            (OPTIONS_NO_ANCHOR, "[$]", '$'..='$', vec![]),
+            (OPTIONS_NO_ANCHOR, "[{$}]", '$'..='$', vec![]),
+            (OPTIONS_ANCHOR, "[$]", '\u{ffff}'..='\u{ffff}', vec![]),
+            (OPTIONS_ANCHOR, r"[\$]", '$'..='$', vec![]),
+            (OPTIONS_ANCHOR, "[{$}]", '$'..='$', vec![]),
+            // TODO: add more tests, look for ICU tests
+        ];
+        for (options, source, single, multi) in cases {
+            let parsed = parse_unstable(source, options, &td!()).unwrap();
+            assert_set_equality(&parsed, single.clone(), multi.iter().map(|s| s.as_str()));
+        }
+    }
+
+    #[test]
+    fn test_error_messages() {
```

Comment:
Nit: You don't test the successful behavior as much. The only tests I see are in docs tests. I think it would be helpful to add more positive tests.

Reply:
Absolutely, filed an issue: #3556

There are a few positive tests in `test_semantics`, but they're by no means exhaustive.